networks/suzuka/suzuka-client/src/bin/e2e/gas_dos.rs

```diff
@@ -0,0 +1,193 @@
+	let timeout_secs = 600; // 10 minutes
+	let max_gas_amount = 5_000;
+	let gas_unit_price = 100;
+
+	let expiration_time =
+		SystemTime::now().duration_since(UNIX_EPOCH).unwrap().as_secs() + timeout_secs;
+
+	let transaction_builder = TransactionBuilder::new(
+		TransactionPayload::EntryFunction(EntryFunction::new(
+			ModuleId::new(AccountAddress::ONE, Identifier::new("coin").unwrap()),
+			Identifier::new("transfer")?,
+			vec![TypeTag::from_str(coin_type)?],
+			vec![to_bytes(&to_account)?, to_bytes(&amount)?],
+		)),
+		expiration_time,
+		ChainId::new(chain_id),
+	);
+
+	let raw_txn = transaction_builder
+		.sender(from_account.address())
+		.sequence_number(from_account.sequence_number())
+		.max_gas_amount(max_gas_amount)
+		.gas_unit_price(gas_unit_price)
+		.expiration_timestamp_secs(expiration_time)
+		.chain_id(ChainId::new(chain_id))
+		.build();
+
+	let signed_txn = from_account.sign_transaction(raw_txn);
+
+	Ok(signed_txn)
+}
+
+pub async fn test_sending_failed_tx() -> Result<(), anyhow::Error> {
+	let rest_client = Client::new(NODE_URL.clone());
+	let faucet_client = FaucetClient::new(FAUCET_URL.clone(), NODE_URL.clone());
+	let coin_client = CoinClient::new(&rest_client);
+
+	let alice = LocalAccount::generate(&mut rand::rngs::OsRng);
+	let bob = LocalAccount::generate(&mut rand::rngs::OsRng);
+
+	println!("=== Addresses ===");
+	println!("Alice: {}", alice.address().to_hex_literal());
+	println!("Bob: {}", bob.address().to_hex_literal());
+
+	faucet_client
+		.fund(alice.address(), 100_000_000)
+		.await
+		.context("Failed to fund Alice's account")?;
+
+	faucet_client
+		.create_account(bob.address())
+		.await
+		.context("Failed to fund Bob's account")?;
+
+	let chain_id = rest_client
+		.get_index()
+		.await
+		.context("Failed to get chain ID")?
+		.inner()
+		.chain_id;
+	println!("\n=== Initial Balance ===");
+	let initial_balance = coin_client
+		.get_account_balance(&alice.address())
+		.await
+		.context("Failed to get Alice's account balance")?;
+	println!("Alice: {:?}", initial_balance);
+
+	// TEST 1: Sending a transaction trying to transfer more coins than Alice has (including gas fees)
+	let transaction =
+		create_fake_signed_transaction(chain_id, &alice, bob.address(), 100_000_000).await?;
+
+	let _transaction_will_fail = rest_client
+		.submit(&transaction)
+		.await
+		.context("Failed when waiting for the transaction")?
+		.into_inner();
+	match rest_client.wait_for_signed_transaction(&transaction).await {
+		Ok(_) => panic!("Transaction should have failed"),
+		Err(e) => {
+			println!("Transaction failed as expected: {:?}", e);
+		}
+	}
+
+	// assert gas fee charged
+	let failed_balance = coin_client
+		.get_account_balance(&alice.address())
+		.await
+		.context("Failed to get Alice's account balance")?;
+	println!("\n=== After Failed Tx#1 ===");
+	println!("Alice: {:?}", failed_balance);
+	assert!(initial_balance > failed_balance);
+
+	// TEST 2: Sending n transactions with a high sequence number
```

Comment:
I see a structural problem here: two tests are rolled into one binary, perhaps because creating binary targets for each test is work-intensive and add a lot boilerplate.

If each of the tests had a `#[test]` entry point, and maybe delegated the setup to a common function, this could be developed as a normal test suite.

Reply:
All of these conditions should hold to prove the `gas_dos` does not occur. But, if you want to adopt a standard of minimization, I can do so. 

I thought we were trying to avoid writing e2e tests under the `#[test]` macro.

Reply:
I'd still prefer `#[test]` entry points because they are easier to maintain in the long run and they plug into the standard `cargo test` tool.

OK with this being a two-step test, though.